### ISSUE TITLE
Modifier key not taken into account when typing to fast

### ISSUE THREAD

**Issue by tchernomax** [ibus/ibus#2008]
# Which distribution and version?: 
Archlinux.

```
max@mde-oxalide % pacman -Qs ibus
local/ibus 1.5.18-1
    Next Generation Input Bus for Linux
…
local/libibus 1.5.18-1
    IBus support library
...
max@mde-oxalide % pacman -Qs gnome-shell
local/gnome-shell 3.28.2-1 (gnome)
    The next generation GNOME Shell
```

# Which desktop environment and version?: 
gnome 3.28

# Which session type?: 
wayland

# Which application and version?: 
gedit 3.28 (and all other gnome applications)

The problem doesn't appear on application using xwayland (eg. firefox & chromium)

# IBus version?: 
IBus 1.5.18

# Issue description:

From gnome-shell 3.28.1 (even if I now use 3.28.2), when I use keyboard macro (send keystroke very quickly) the modifier keys aren't taken into account.

"Shift+," lead to "," instead of "<"
```
Event: time 1523785942.236209, type 4 (EV_MSC), code 4 (MSC_SCAN), value 700e1
Event: time 1523785942.236209, type 1 (EV_KEY), code 42 (KEY_LEFTSHIFT), value 1
Event: time 1523785942.236209, -------------- SYN_REPORT ------------
Event: time 1523785942.237034, type 4 (EV_MSC), code 4 (MSC_SCAN), value 70036
Event: time 1523785942.237034, type 1 (EV_KEY), code 51 (KEY_COMMA), value 1
Event: time 1523785942.237034, -------------- SYN_REPORT ------------
Event: time 1523785942.238016, type 4 (EV_MSC), code 4 (MSC_SCAN), value 70036
Event: time 1523785942.238016, type 1 (EV_KEY), code 51 (KEY_COMMA), value 0
Event: time 1523785942.238016, -------------- SYN_REPORT ------------
Event: time 1523785942.239014, type 4 (EV_MSC), code 4 (MSC_SCAN), value 700e1
Event: time 1523785942.239014, type 1 (EV_KEY), code 42 (KEY_LEFTSHIFT), value 0
Event: time 1523785942.239014, -------------- SYN_REPORT ------------
,
```

I git bisect the regression on gnome-shell to the commit [36c7d65c](https://gitlab.gnome.org/GNOME/gnome-shell/commit/36c7d65ccf7d3927dc9b6a9356291b92ae249c94).

```
commit 36c7d65ccf7d3927dc9b6a9356291b92ae249c94
Author: Carlos Garnacho <carlosg@gnome.org>
Date:   Thu Mar 15 14:01:51 2018 +0100

    inputMethod: Pass all key events through the current IM method

    Even though we are using an "xkb" source, it still makes sense to
    pass the event through the IBus simple engine, in order to let it
    handle compose keys and ctrl+shift+[u|e].

    https://gitlab.gnome.org/GNOME/gnome-shell/issues/115

    Closes: #115

diff --git a/js/misc/inputMethod.js b/js/misc/inputMethod.js
index addc20f4e..621483243 100644
--- a/js/misc/inputMethod.js
+++ b/js/misc/inputMethod.js
@@ -188,8 +188,7 @@ var InputMethod = new Lang.Class({
     vfunc_filter_key_event(event) {
         if (!this._context || !this._enabled)
             return false;
-        if (!this._currentSource ||
-            this._currentSource.type == Keyboard.INPUT_SOURCE_TYPE_XKB)
+        if (!this._currentSource)
             return false;

         let state = event.get_state();
```

With gnome-shell 3.28.0 or 3.28.1 with 36c7d65c reverted:

```
Event: time 1523786131.349189, type 4 (EV_MSC), code 4 (MSC_SCAN), value 700e1
Event: time 1523786131.349189, type 1 (EV_KEY), code 42 (KEY_LEFTSHIFT), value 1
Event: time 1523786131.349189, -------------- SYN_REPORT ------------
Event: time 1523786131.350028, type 4 (EV_MSC), code 4 (MSC_SCAN), value 70036
Event: time 1523786131.350028, type 1 (EV_KEY), code 51 (KEY_COMMA), value 1
Event: time 1523786131.350028, -------------- SYN_REPORT ------------
Event: time 1523786131.351015, type 4 (EV_MSC), code 4 (MSC_SCAN), value 70036
Event: time 1523786131.351015, type 1 (EV_KEY), code 51 (KEY_COMMA), value 0
Event: time 1523786131.351015, -------------- SYN_REPORT ------------
Event: time 1523786131.351951, type 4 (EV_MSC), code 4 (MSC_SCAN), value 700e1
Event: time 1523786131.351951, type 1 (EV_KEY), code 42 (KEY_LEFTSHIFT), value 0
Event: time 1523786131.351951, -------------- SYN_REPORT ------------
<
```

I filled a issue on the gnome-shell gitlab : https://gitlab.gnome.org/GNOME/gnome-shell/issues/208
But [Carlos Garnacho told me it is probably a ibus issue](https://gitlab.gnome.org/GNOME/gnome-shell/issues/208#note_97176), so here I am :)

# Steps to reproduce:

You have to type extremely fast or use a keyboard that allow you to define macros.

1. Type "Shift+,"

You will see a "," instead of a "<".

# Can you reproduce your problem when you restart ibus-daemon? (yes / no): 

very interestingly no

```
max@mde-oxalide % u2026
max@mde-oxalide % ps -ef | grep ibus
gdm        829   706  0 13:16 tty1     00:00:00 ibus-daemon --xim --panel disable
gdm        832   829  0 13:16 tty1     00:00:00 /usr/lib/ibus/ibus-dconf
gdm        835     1  0 13:16 tty1     00:00:00 /usr/lib/ibus/ibus-x11 --kill-daemon
gdm        840   678  0 13:16 ?        00:00:00 /usr/lib/ibus/ibus-portal
gdm        902   829  0 13:16 tty1     00:00:00 /usr/lib/ibus/ibus-engine-simple
max       1071  1017  0 13:17 tty2     00:00:00 ibus-daemon --xim --panel disable
max       1075  1071  0 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-dconf
max       1076  1071  2 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-extension-gtk3
max       1080     1  0 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-x11 --kill-daemon
max       1084   966  0 13:17 ?        00:00:00 /usr/lib/ibus/ibus-portal
max       1243  1071  0 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-engine-simple
max       1739  1657  0 13:17 pts/1    00:00:00 grep --color ibus
max@mde-oxalide % ibus exit
max@mde-oxalide % ps -ef | grep ibus
gdm        829   706  0 13:16 tty1     00:00:00 ibus-daemon --xim --panel disable
gdm        832   829  0 13:16 tty1     00:00:00 /usr/lib/ibus/ibus-dconf
gdm        835     1  0 13:16 tty1     00:00:00 /usr/lib/ibus/ibus-x11 --kill-daemon
gdm        840   678  0 13:16 ?        00:00:00 /usr/lib/ibus/ibus-portal
gdm        902   829  0 13:16 tty1     00:00:00 /usr/lib/ibus/ibus-engine-simple
max       1745  1017  0 13:17 tty2     00:00:00 ibus-daemon --xim --panel disable
max       1750  1745  0 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-dconf
max       1751  1745  7 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-extension-gtk3
max       1753     1  1 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-x11 --kill-daemon
max       1756   966  0 13:17 ?        00:00:00 /usr/lib/ibus/ibus-portal
max       1770  1745  0 13:17 tty2     00:00:00 /usr/lib/ibus/ibus-engine-simple
max       1779  1657  0 13:17 pts/1    00:00:00 grep --color ibus
max@mde-oxalide % …
```

note: where you see "u2026 " and "…" I typed the same thing : ctrl+shift+u 2 0 2 6 space

# Do you see any errors when you run ibus-daemon with the verbose option?:

gnome restart automatically ibus when I kill it.

# Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 

yes

**Comment by fujiwarat** [comment]
I tried to type `Shift-,` quickly but I cannot reproduce your problem in gnome-shell 3.28.2.
Since I don't have the programmable keyboard, I think it's good to simulate key events with API but I don't know how to simulate the key events in Wayland while I can do with XTest in Xorg.

My understanding is your problem is not reproduced in Xorg and I have no idea to investigate your problem.

> gnome restart automatically ibus when I kill it.

The easy way is to rename ibus-daemon not to restart it by gnome-shell.
```
% sudo /usr/bin/ibus-daemon /usr/bin/ibus-daemon.orig
% ibus exit
% ibus-daemon.orig --verbose --xim --panel disable &
```

**Comment by tchernomax** [comment]
I checked the environment variables:
```
% ps -ef | grep ibus-daemon
gdm        816   692  0 20:29 tty1     00:00:00 /bin/sh /usr/bin/ibus-daemon --xim --panel disable
max       2587  1029  0 20:37 tty2     00:00:00 ibus-daemon --xim --panel disable
max       2630  2507  0 20:38 pts/3    00:00:00 grep --color ibus-daemon
% sudo strings /proc/1029/environ | sort > env1
% ibus exit
% ps -ef | grep ibus-daemon
gdm        816   692  0 20:29 tty1     00:00:00 /bin/sh /usr/bin/ibus-daemon --xim --panel disable
max       2694  1029  0 20:41 tty2     00:00:00 ibus-daemon --xim --panel disable
max       2726  2507  0 20:41 pts/3    00:00:00 grep --color ibus-daemon 
% sudo strings /proc/2694/environ | sort > env2
% diff env1 env2
2d1
< DESKTOP_AUTOSTART_ID=10815919c288f3fa8d152632229225067500000010060000
```

I also tried to launch ibus-daemon with verbose:
```
% sudo /usr/bin/ibus-daemon /usr/bin/ibus-daemon.orig
% ibus exit
% ibus-daemon.orig --verbose --xim --panel disable
```
The output stay blank

I also did:
```
max@mde-oxalide % sudo mv /usr/bin/ibus-daemon{,-orig}
max@mde-oxalide % sudo vim /usr/bin/ibus-daemon
max@mde-oxalide % cat /usr/bin/ibus-daemon
#!/bin/sh
/usr/bin/ibus-daemon-orig --verbose "$@" &> /tmp/ibus.log
max@mde-oxalide % sudo reboot
…
max@mde-oxalide % cat /tmp/ibus.log
max@mde-oxalide % pgrep -a ibus-daemon
837 /bin/sh /usr/bin/ibus-daemon --xim --panel disable
838 /usr/bin/ibus-daemon-orig --verbose --xim --panel disable
max@mde-oxalide % sudo kill 838
max@mde-oxalide % pgrep -a ibus-daemon
1938 /bin/sh /usr/bin/ibus-daemon --xim --panel disable
1939 /usr/bin/ibus-daemon-orig --verbose --xim --panel disable
max@mde-oxalide % cat /tmp/ibus.log
max@mde-oxalide %
```
The ibus-daemon of the session isn't launched, only the ibus-daemon of gdm.
Without the session ibus-daemon, the problem doesn't append.

**Comment by fujiwarat** [comment]
And then I have no idea to simulate and evaluate your problem.

I uploaded one [ibus-keypress.c](https://github.com/ibus/ibus/blob/master/src/tests/ibus-keypress.c).
```
% env GTK_IM_MODULE=ibus ./ibus-keypress
```
But this works with GNOME Xorg or other desktops but not GNOME Wayland.

**Comment by egalanos** [comment]
I hit this same issue today when using both a Mooltipass and a yubikey (static password containing shifted keys).

An ibus restart also made it go away (for now)

**Comment by egalanos** [comment]
I'm able to reproduce it 100% on Fedora 28 with all updates installed by playing back this recorded input from libinput
[libinput-fast.yaml.gz](https://github.com/ibus/ibus/files/2281126/libinput-fast.yaml.gz)

1. Open a gnome-terminal and run:

```
cat > /dev/null
```

2. Open another gnome-terminal. As **root** run:

```
sleep 5;yes '\n' | libinput replay libinput-fast.yaml 
```

3. Quickly move keyboard focus to the terminal redirecting to /dev/null.

It will be being sent a continuous output of _AbCdEfGhIjKlMnOpQrStUvWxYz_

Any output other than that is an indicator of the bug.

4. Run ibus restart

Problem will then be gone until next reboot.

**Comment by fujiwarat** [comment]
> I'm able to reproduce it 100% on Fedora 28 with all updates installed by playing back this recorded input from libinput

It seems works fine from the [log](https://github.com/ibus/ibus/files/2281887/libinput-fujiwara.txt)

**Comment by fujiwarat** [comment]
I think the original issue is wayland specific.

**Comment by egalanos** [comment]
My Fedora 28 desktop is also Gnome on Wayland like the original report.

It didn't show up in gvim (Xwayland rather than wayland).

**Comment by tchernomax** [comment]
The problem still append in gnome 3.32 (still wayland).
It's even worse as now I have to make /usr/bin/ibus-daemon inaccessible (`ibus exit` after boot doesn't solve the problem anymore).

**Comment by fujiwarat** [comment]
How about the latest gnome 3.30.y instead of 3.32? I have a know issue in 3.32 about key typing.

**Comment by tchernomax** [comment]
@fujiwarat when I was using gnome 3.30.2 the problem was present but `ibus exit` was enough to fix it for the session time.

**Comment by fujiwarat** [comment]
OK, unfortunately I cannot reproduce your problem.
I guess your problem is not always reproduced.

**Comment by tchernomax** [comment]
@fujiwarat 
> I guess your problem is not always reproduced.
The problem always reproduce (as long as I am on Wayland with ibus running).

You can try it with the file supplied by @egalanos 

For example, on Xorg with ibus running or on Wayland without ibus :
```
max@mde-oxalide % sleep 5 ; echo -e "\n\n\n" | sudo libinput replay libinput-fast.yaml
/dev/input/event25: SE Mooltipass
Hit enter to start replayingHit enter to start replayingHit enter to start replayingHit enter to start replayingHit enter to start replayingTraceback (most recent call last):
  File "/usr/lib/libinput/libinput-replay", line 207, in <module>
    main()
  File "/usr/lib/libinput/libinput-replay", line 197, in main
    loop(args, y)
  File "/usr/lib/libinput/libinput-replay", line 169, in loop
    input('Hit enter to start replaying')
EOFError: EOF when reading a line

(in parallel on another terminal)
max@mde-oxalide % cat > /dev/null
nOpQrStUvWxYz
AbCdEfGhIjKlMnOpQrStUvWxYz
AbCdEfGhIjKlMnOpQrStUvWxYz
AbCdEfGhIjKlMnOpQrStUvWxYz
```

Don't mind the first line (nOpQrStUvWxYz) it's irrelevant ; you can see that all the lines are equal → that's the expected behavior.

On Wayland with ibus running:
```
max@mde-oxalide % sleep 5 ; echo -e "\n\n\n" | sudo libinput replay libinput-fast.yaml
/dev/input/event25: SE Mooltipass
Hit enter to start replayingHit enter to start replayingHit enter to start replayingHit enter to start replayingHit enter to start replayingTraceback (most recent call last):
  File "/usr/lib/libinput/libinput-replay", line 207, in <module>
    main()
  File "/usr/lib/libinput/libinput-replay", line 197, in main
    loop(args, y)
  File "/usr/lib/libinput/libinput-replay", line 169, in loop
    input('Hit enter to start replaying')
EOFError: EOF when reading a line

(in parallel on another terminal)
max@mde-oxalide % cat > /dev/null
pqrstUvwxyz
abcdefghijklmnopqrstuvwxyz
abCDEFGHIJKLMNOPQrstuvwXyz
ABCDeFGHIjklMNOPQRSTuvwxyz
```

Don't mind the first line it's irrelevant ; you can see that all the lines **are not** equal → they should

**Comment by fujiwarat** [comment]
@ tchernomax I think your problem is not reproduced in Xorg and it's not good to use that script.

**Comment by tchernomax** [comment]
@fujiwarat 

> I think your problem is not reproduced in Xorg

Yes, the problemn is wayland specific, as I said in my previous post:

> on Xorg with ibus running or on Wayland without ibus [...] you can see that all the lines are equal → that's the expected behavior.
> On Wayland with ibus running [...] you can see that all the lines are not equal → they should

But now wayland is the default display server for gnome.

> it's not good to use that script.

I don't understand what you mean. `libinput replay` reproduce exactly what I get when I use my programmable keyboard, or any device that produce fast keystrokes.

**Comment by fujiwarat** [comment]
> I don't understand what you mean. libinput replay reproduce exactly what I get when I use my programmable keyboard, or any device that produce fast keystrokes.

I don't think your problem would be the same root cause with the script issue. Regarding to the script issue, I guess the script could add an idle time likes g_idle_add() and it would be a low priority for me.

**Comment by tchernomax** [comment]
@fujiwarat I do think it's because there is no idle time between keystrokes that the problem happen on my keyboardio and on yubikeys (or every device that simulate keystrokes).

Yes the script could be "fixed" by adding an idle time between keystroke… but that's exactly my point → ibus can't currently handle rapid keystrokes (keystrokes with almost zero idle time between them).

**Comment by fujiwarat** [comment]
> I do think it's because there is no idle time between keystrokes that the problem happen on my keyboardio and on yubikeys (or every device that simulate keystrokes).

And then probably I'd think your problem would be different from the script issue and I need to reproduce your problem to investigate what happens.

**Comment by georgmu** [comment]
I have the same issue with a barcode scanner, where the time between "Shift" and the next key is very short. The shift press is often, but not always, not recognised. "ibus exit" sometimes fixes it, sometimes not. I am using Fedora 30 with Wayland desktop. Going to a tty fixes it.

Sometimes it even modifies the character before:
A sequence which should result in `84X9` gets printed as `8$X9` with the following libinput record: 
```
  - evdev:
    - [ 12, 888260,   4,   4, 458789] # EV_MSC / MSC_SCAN             458789
    - [ 12, 888260,   1,   9,      1] # EV_KEY / KEY_8                     1
    - [ 12, 888260,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms
  - evdev:
    - [ 12, 896280,   4,   4, 458789] # EV_MSC / MSC_SCAN             458789
    - [ 12, 896280,   1,   9,      0] # EV_KEY / KEY_8                     0
    - [ 12, 896280,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms
  - evdev:
    - [ 12, 904272,   4,   4, 458785] # EV_MSC / MSC_SCAN             458785
    - [ 12, 904272,   1,   5,      1] # EV_KEY / KEY_4                     1
    - [ 12, 904272,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms
  - evdev:
    - [ 12, 912264,   4,   4, 458785] # EV_MSC / MSC_SCAN             458785
    - [ 12, 912264,   1,   5,      0] # EV_KEY / KEY_4                     0
    - [ 12, 912264,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms
  - evdev:
    - [ 12, 920263,   4,   4, 458977] # EV_MSC / MSC_SCAN             458977
    - [ 12, 920263,   1,  42,      1] # EV_KEY / KEY_LEFTSHIFT             1
    - [ 12, 920263,   4,   4, 458779] # EV_MSC / MSC_SCAN             458779
    - [ 12, 920263,   1,  45,      1] # EV_KEY / KEY_X                     1
    - [ 12, 920263,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms
  - evdev:
    - [ 12, 928300,   4,   4, 458977] # EV_MSC / MSC_SCAN             458977
    - [ 12, 928300,   1,  42,      0] # EV_KEY / KEY_LEFTSHIFT             0
    - [ 12, 928300,   4,   4, 458779] # EV_MSC / MSC_SCAN             458779
    - [ 12, 928300,   1,  45,      0] # EV_KEY / KEY_X                     0
    - [ 12, 928300,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms
  - evdev:
    - [ 12, 936277,   4,   4, 458790] # EV_MSC / MSC_SCAN             458790
    - [ 12, 936277,   1,  10,      1] # EV_KEY / KEY_9                     1
    - [ 12, 936277,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms
  - evdev:
    - [ 12, 944308,   4,   4, 458790] # EV_MSC / MSC_SCAN             458790
    - [ 12, 944308,   1,  10,      0] # EV_KEY / KEY_9                     0
    - [ 12, 944308,   0,   0,      0] # ------------ SYN_REPORT (0) ---------- +8ms

```

When doing multiple scans, I see `84x9` more often than this, but the correct result (`84X9`) is very rare.

**Comment by tchernomax** [comment]
@georgmu `ibus exit` doesn't work anymore, gnome restart ibus automatically. The current working workaround I use `sudo mv /usr/bin/ibus-daemon{,-orig}`.

**Comment by georgmu** [comment]
@tchernomax thx, but I would like to have a proper solution without any workarounds

Is there some overview which components are involved from start (= something happens in /dev/input/) to the end (= a character appears on the display.

One of the components in between (ibus or wayland or gtk or ...) messes up, does some handling asynchronous even for the same input source (I can't imagine how the otherwise correct synchronous order of libinput gets wrong).

**Comment by tchernomax** [comment]
> @tchernomax thx, but I would like to have a proper solution without any workarounds

I can't agree more ! Yet I didn't found the proper solution for this bug.

> Is there some overview which components are involved from start (= something happens in /dev/input/) to the end (= a character appears on the display.

An overview would be indeed a good debug starting point.

**Comment by georgmu** [comment]
I did another check using WAYLAND_DEBUG=1 with a gedit window:

Input is always `84X9`.

Correct sequence with `84X9` as output:
```
[1008402,794] wl_keyboard@28.key(47633, 19941114, 9, 1)
[1008402,823] wl_keyboard@28.key(47634, 19941122, 9, 0)
[1008402,847] wl_keyboard@28.key(47635, 19941130, 5, 1)
[1008402,890] wl_keyboard@28.modifiers(47636, 1, 0, 0, 0)
[1008419,636] wl_keyboard@28.key(47637, 19941138, 5, 0)
[1008419,676] wl_keyboard@28.key(47638, 19941146, 42, 1)
[1008419,702] wl_keyboard@28.key(47639, 19941146, 45, 1)
[1008419,730] wl_keyboard@28.modifiers(47640, 0, 0, 0, 0)
[1008435,688] wl_keyboard@28.key(47641, 19941154, 42, 0)
[1008435,744] wl_keyboard@28.key(47642, 19941154, 45, 0)
[1008435,769] wl_keyboard@28.key(47643, 19941162, 10, 1)
[1008461,574] wl_keyboard@28.key(47644, 19941170, 10, 0)
```
(even here the modifiers fires one event too early, but this resulted in no issue, because it was after the press)

Key map is '8' = 9, '4' = 5, 'Shift' = 42, 'X' = 45.
Last number in the key events is 1 = press, 0 = release.

The following sequence ended up as `8$x9`:
```
[691425,425] wl_keyboard@28.key(46311, 19624145, 9, 1)
[691425,479] wl_keyboard@28.key(46312, 19624153, 9, 0)
[691445,145] wl_keyboard@28.modifiers(46313, 1, 0, 0, 0)
[691445,187] wl_keyboard@28.key(46314, 19624161, 5, 1)
[691445,230] wl_keyboard@28.key(46315, 19624169, 5, 0)
[691445,259] wl_keyboard@28.modifiers(46316, 0, 0, 0, 0)
[691459,723] wl_keyboard@28.key(46317, 19624177, 42, 1)
[691459,760] wl_keyboard@28.key(46318, 19624177, 45, 1)
[691459,782] wl_keyboard@28.key(46319, 19624185, 42, 0)
[691459,809] wl_keyboard@28.key(46320, 19624185, 45, 0)
[691476,507] wl_keyboard@28.key(46321, 19624193, 10, 1)
[691476,563] wl_keyboard@28.key(46322, 19624201, 10, 0)
```

So the Events itself are all in order, but the modifiers event is fired far too early.

**Comment by georgmu** [comment]
I did another run with ibus disbaled (as suggested by @tchernomax).

In this case, the modifiers event always fires at the right place, right after the change of the shift key press/release (42):
```
[3127538,927] wl_keyboard@28.key(60843, 29931134, 9, 1)
[3127539,017] wl_keyboard@28.key(60844, 29931142, 9, 0)
[3127539,046] wl_keyboard@28.key(60845, 29931150, 5, 1)
[3127539,069] wl_keyboard@28.key(60846, 29931158, 5, 0)
[3127559,949] wl_keyboard@28.key(60847, 29931166, 42, 1)
[3127560,035] wl_keyboard@28.modifiers(60848, 1, 0, 0, 0)
[3127560,093] wl_keyboard@28.key(60849, 29931166, 45, 1)
[3127560,153] wl_keyboard@28.key(60850, 29931174, 42, 0)
[3127560,202] wl_keyboard@28.modifiers(60851, 0, 0, 0, 0)
[3127560,244] wl_keyboard@28.key(60852, 29931174, 45, 0)
[3127578,943] wl_keyboard@28.key(60853, 29931182, 10, 1)
[3127578,999] wl_keyboard@28.key(60854, 29931190, 10, 0)
```

**Comment by luongthanhlam** [comment]
@fujiwarat  I have the same problem too, not only modifier key but also backspace key. You can reproduce this by running the script below on firefox with and without ibus's running
https://gist.github.com/luongthanhlam/52440ade4bab3f162f34e63c1c5979eb

**Comment by tchernomax** [comment]
this bug is still present in 1.5.21

**Comment by tchernomax** [comment]
The problem is still present in gnome 3.34.1
